File: Algorithms/Dynamic_Programming/Egg_Drop/egg_drop.py
INT_MAX=32767

def egg_drop(n,k):
    m=[[0 for x in range(k+1)]for x in range(n+1)]
    for i in range(1,n+1):
        m[i][0]=0
        m[i][1]=1
    
    for j in range(1,k+1):
        m[1][j]=j

    for i in range(2,n+1):
        for j in range(2,k+1):
            m[i][j]=INT_MAX
            for x in range(1,j+1):
               res = 1 + max(m[i-1][x-1], m[i][j-x]) 
               if res < m[i][j]: 
                m[i][j] = res
    print(m) 
    return m[n][k]

n=2
k=6
print(egg_drop(n,k))

File: Algorithms/Dynamic-Programming/Egg_Drop/egg_drop.ts
const eggDrop = (n: number, k: number) => {
  let eggFloor: number[][] = [];

  // A 2d array where entry eggFloor[i][j] will represent
  // the minimum number of trials needeed for i eggs and j floors
  for (let i = 0; i < n + 1; i++) {
    eggFloor.push(new Array(k + 1).fill(0));
  }

  // One trial for one floor and 0 trial for zero floor
  for (let i = 1; i < n + 1; i++) {
    eggFloor[i][1] = 1;
    eggFloor[i][0] = 0;
  }

  // We always need j trials for one egg and j floors
  for (let j = 1; j < k + 1; j++) {
    eggFloor[1][j] = j;
  }

  // Fill the rest of the entries in table using optimal
  // subtructure property
  for (let i = 2; i < n + 1; i++) {
    for (let j = 2; j < k + 1; j++) {
      eggFloor[i][j] = Number.MAX_SAFE_INTEGER;
      for (let x = 0; x < j + 1; x++) {
        let res = 1 + Math.max(eggFloor[i - 1][x - 1], eggFloor[i][j - x]);
        if (res < eggFloor[i][j]) {
          eggFloor[i][j] = res;
        }
      }
    }
  }

  // eggFloor[n][k] holds the result
  return eggFloor[n][k];
};

const n = 2;
const k = 36;
console.log(
  "Minimum worst case of trials in worst case with " +
    n +
    " eggs " +
    k +
    " floors is " +
    eggDrop(n, k)
);
